Fix several glob/path bugs

dtivel opened the pull request:
Fix NuGet/Home#3526.
Fix NuGet/Home#3405.
Fix a bug with `PathResolver.NormalizeBasePath(...)` where `../` was unsupported on Unix-flavored operating systems.
Add >150 unit tests.
Add copyright headers to files.

@rrelyea @emgarten @joelverhagen @drewgil @alpaix @mishra14 @jainaashish @zhili1208

dtivel committed:
Fix several glob/path bugs

Fix NuGet/Home#3526.
Fix NuGet/Home#3405.
Fix a bug with PathResolver.NormalizeBasePath(...) where ../ was
unsupported on Unix-flavored operating systems.
Add >150 unit tests.
Add copyright headers to files.

dtivel committed:
Remove superfluous verbatim string literal

dnfclas commented:
Hi __@dtivel__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

emgarten commented on src/NuGet.Core/NuGet.Common/PathUtil/PathResolver.cs:
```diff
@@ -173,17 +176,18 @@ internal static string GetPathToEnumerateFrom(string basePath, string searchPath
+            // If the search path is relative, transfer the parent directory portion to the base path.
             // This needs to be done because the base path determines the root for our enumeration.
-            while (searchPath.StartsWith(relativePath, StringComparison.OrdinalIgnoreCase))
+            while (searchPath.StartsWith(parentDirectoryPath, StringComparison.OrdinalIgnoreCase))
```
Is IgnoreCase safe for case sensitive file systems here?

dtivel replied:
In this specific case I think it's benign, since the "search for" strings `../` and `..\` are unaffected by casing.

emgarten reviewed (approved):
Great tests! 👏 :shipit:

jainaashish reviewed (approved):
cool, nice set of tests!

dtivel committed:
Sort expected and actual results

Some tests fail on the CI build's macOS and Linux machines because sort
ordering was unexpected.  Oddly, the same tests pass on my macOS
machine.

joelverhagen commented on test/NuGet.Core.Tests/NuGet.Packaging.Test/PackageBuilderTest.cs:
```diff
@@ -2603,5 +2639,65 @@ private Stream GetManifestStream(Stream packageStream)
+            file1.txt
+            file2.txt
+        */
+        public sealed class TestDirectory : IDisposable
```
Can this share some implementation with the existing `TestDirectory`?

joelverhagen commented on test/NuGet.Core.Tests/NuGet.Common.Test/PathResolverTests.cs:
```diff
@@ -0,0 +1,703 @@
+    */
+    public sealed class TestDirectoryFixture : IDisposable
+    {
+        private DirectoryInfo _directory;
```
Same as below, can this share some implementation with either the existing `TestDirectory` or the other `TestDirectory` below?

dtivel replied:
I'll do so in a later PR.

joelverhagen commented on test/NuGet.Core.Tests/NuGet.Common.Test/PathResolverTests.cs:
```diff
@@ -0,0 +1,703 @@
+                }, actualFullPaths);
+        }
+
+        [Platform(Platform.Darwin)]
```
This looks the same as the Linux test, can a `[Theory]` have two `[Platform]` attributes to act as an `or`?

dtivel replied:
No, the Platform attribute is unique.

dtivel replied:
To elaborate, the `PlatformAttribute` attribute [does not allow multiple decorations](https://github.com/NuGet/NuGet.Client/blob/bb7f380e7e9c87768d0b562488990becfddfcb78/src/NuGet.Core/NuGet.Test.Utility/Traits/PlatformAttribute.cs#L13).

Changing `AllowMultiple` to true does not solve this, since `xunit.console.exe -notrait` excludes tests with the specified trait.  This is why I have platform-specific versions of some tests.

joelverhagen replied:
Cool, thanks.
